File: cropguard-ai/backend/app/api/fields.py
from fastapi import APIRouter, Depends, HTTPException

from app.schemas.fields import FieldCreate, FieldOut
from app.services.mongo_client import DatabaseError, db_insert, db_select
from app.utils.auth import get_current_user

router = APIRouter(prefix="/api/fields", tags=["fields"])


def _row_to_field(row: dict) -> FieldOut:
    return FieldOut(
        id=row["id"],
        name=row["name"],
        crop=row["crop"],
        area_acres=row["area"],
        latitude=row["latitude"],
        longitude=row["longitude"],
        health_status=row["health_status"],
        last_scan_at=row.get("last_scan_at"),
        last_disease=row.get("last_disease"),
        last_risk=row.get("last_risk"),
    )


@router.get("", response_model=list[FieldOut])
async def list_fields(user: dict = Depends(get_current_user)):
    try:
        rows = await db_select(
            "fields", {"user_id": f"eq.{user['id']}", "order": "created_at.desc", "select": "*"}
        )
    except DatabaseError as e:
        raise HTTPException(e.status_code, str(e))
    return [_row_to_field(r) for r in rows]


@router.get("/{field_id}", response_model=FieldOut)
async def get_field(field_id: str, user: dict = Depends(get_current_user)):
    try:
        rows = await db_select(
            "fields", {"id": f"eq.{field_id}", "user_id": f"eq.{user['id']}", "select": "*"}
        )
    except DatabaseError as e:
        raise HTTPException(e.status_code, str(e))
    if not rows:
        raise HTTPException(404, "Field not found.")
    return _row_to_field(rows[0])


@router.post("", response_model=FieldOut)
async def create_field(payload: FieldCreate, user: dict = Depends(get_current_user)):
    try:
        row = await db_insert(
            "fields",
            {
                "user_id": user["id"],
                "name": payload.name,
                "crop": payload.crop,
                "area": payload.area_acres,
                "latitude": payload.latitude,
                "longitude": payload.longitude,
                "health_status": "healthy",
            },
        )
    except DatabaseError as e:
        raise HTTPException(e.status_code, str(e))
    return _row_to_field(row)

File: cropguard-ai/frontend/lib/api/fields.ts
import { apiFetch, ApiError, mockDelay, USE_MOCK_API } from "./client";
import { mockFields } from "@/lib/mock-data";
import type { Field } from "@/types";

// In-memory store so additions persist for the session (mock only —
// resets on page reload; Stage 3 replaces this with Supabase).
let fieldsStore: Field[] = [...mockFields];

// Backend rows are snake_case (see backend/app/schemas/fields.py); the
// frontend's Field type is camelCase. This is the one place that translates
// between them so every page/component can stay on the camelCase shape
// regardless of which mode (mock or real) is active.
function fromApiRow(row: any): Field {
  return {
    id: row.id,
    name: row.name,
    crop: row.crop,
    areaAcres: row.area_acres,
    latitude: row.latitude,
    longitude: row.longitude,
    healthStatus: row.health_status,
    lastScanAt: row.last_scan_at ?? null,
    lastDisease: row.last_disease ?? null,
    lastRisk: row.last_risk ?? null,
  };
}

export async function listFields(): Promise<Field[]> {
  if (!USE_MOCK_API) {
    const rows = await apiFetch<any[]>("/api/fields");
    return rows.map(fromApiRow);
  }
  await mockDelay(500);
  return fieldsStore;
}

export async function getField(id: string): Promise<Field> {
  if (!USE_MOCK_API) {
    const row = await apiFetch<any>(`/api/fields/${id}`);
    return fromApiRow(row);
  }
  await mockDelay(400);
  const field = fieldsStore.find((f) => f.id === id);
  if (!field) throw new ApiError("Field not found.", 404);
  return field;
}

export interface CreateFieldPayload {
  name: string;
  crop: string;
  areaAcres: number;
  latitude: number;
  longitude: number;
}

export async function createField(payload: CreateFieldPayload): Promise<Field> {
  if (!USE_MOCK_API) {
    const row = await apiFetch<any>("/api/fields", {
      method: "POST",
      body: JSON.stringify({
        name: payload.name,
        crop: payload.crop,
        area_acres: payload.areaAcres,
        latitude: payload.latitude,
        longitude: payload.longitude,
      }),
    });
    return fromApiRow(row);
  }
  await mockDelay(700);
  if (!payload.name.trim()) throw new ApiError("Field name is required.", 400);
  const newField: Field = {
    id: `field-${Date.now()}`,
    name: payload.name,
    crop: payload.crop,
    areaAcres: payload.areaAcres,
    latitude: payload.latitude,
    longitude: payload.longitude,
    healthStatus: "healthy",
    lastScanAt: null,
    lastDisease: null,
    lastRisk: null,
  };
  fieldsStore = [newField, ...fieldsStore];
  return newField;
}
